sound/soc/sof/topology.c

```diff
@@ -3619,7 +2025,21 @@ static int sof_complete(struct snd_soc_component *scomp)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
 	struct snd_sof_widget *swidget, *comp_swidget;
-	int ret;
+	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
+	const struct ipc_tplg_widget_ops *widget_ops = ipc_tplg_ops->widget;
+	int ret = 0;
+
+	/* now update all widget IPC structures */
+	list_for_each_entry(swidget, &sdev->widget_list, list) {
+		if (widget_ops[swidget->id].ipc_setup) {
+			ret = widget_ops[swidget->id].ipc_setup(scomp, swidget);
+			if (ret < 0) {
+				dev_err(sdev->dev, "failed updating IPC struct for %s\n",
+					swidget->widget->name);
+				return ret;
```

Comment:
We don't need to roll back anything, right?

Reply:
best to add a comment to keep track of the behavior